### ISSUE TITLE
Spanish videos missing from kalite 0.15.1

### ISSUE THREAD

**Issue by needlestack** [rachelproject/rachelpiOS#21]
As discussed here, kalite 0.15.1 doesn't have a complete spanish video collection - that is: there are missing videos when browsing topics, but you can't download those videos through the kalite interface. Even if you download them yourself and put them in the content directory, they still don't work.

This seems to be missing info in `/root/.kalite/locale/dubbed_video_mappings.json`

With this corrected there are still problems with caching and a sparse topic tree (lots of topics have not been translated into Spanish). We need to trim the topic tree and get caching working.

**Comment by needlestack** [comment]
Ok, here's our final options for kalite, neither perfect:
1. turn off caching - 10 minutes startup every time - everything beyond that works nicely, though - the topic tree is trimmed to just the well-covered subjects, nothing is grayed out.
2. turn on caching - 30 second startup but
     a) we can't trim the topic tree, so all the topics show and there are many grayed out branches (though all the same Spanish content is there)
     b) the text on the video pages is in English. The videos are fine, it's just the title and description. Everything else is in Spanish.

Current build goes with 1. I've submitted [an issue to the kalite team about how caching causes problems with Spanish videos displaying](https://github.com/learningequality/ka-lite/issues/5066)

Here are the files that need to be swapped out after you've got your kalite installed via the standard rachelpiOS install:
- [dubbed_video_mappings.json.zip](https://github.com/rachelproject/rachelpiOS/files/199744/dubbed_video_mappings.json.zip)
  This goes in the /root/.kalite/locale folder, replacing the file that is there. This includes the mapping info (english ID to spanish ID) for the 200+ videos that have been translated since 0.15 was released
- [topics.json.zip](https://github.com/rachelproject/rachelpiOS/files/199745/topics.json.zip)
  This goes in the /usr/share/kalite/data/khan folder, replacing the file that is there. This is a trimmed version of the topic tree that excludes several thousand topics that do not have a Spanish translation.
- installer.py (in github)
  I've just updated this file to turn off caching again, so that's how it will install. This is really just an issue for the Spanish version, but for now I wanted our builds to work as described in #1 above. If you've already installed with the previous version, but you want the trimmed topic tree to work, you have to edit `/root/.kalite/settings.py` and comment out the line that says `from kalite.project.settings.raspberry_pi import *` to disable caching.

So there we are. Even without the videos you can check out the new trimmed topic tree and see if there's anything else you want to discard for space.

I learned a lot about kalite through this, and I'll learn more in the future hopefully figuring out how to get around these problems. But it won't happen in time for the Spanish release.

I'll have limited connectivity over the weekend, but will keep checking in!

**Comment by julianharty** [comment]
Thanks so much for the detailed investigation. I hope it'll help me get ka-lite working with Swahili contents too (which nearly work OK when I tested downloading them recently).

**Comment by needlestack** [comment]
Let's hope! The two files mentioned above, being JSON format, are fairly easy to work with. You'll have to edit them to your liking as I only tweaked them for Spanish content.

I'm not totally sure of this, but it seemed that the `dubbed_video_mappings.json` file needed to be kept in order. Even though the format is a hash table (and thus technically has no determined order) when I changed the order of languages it seemed to stop working. So I had to manually edit my changes into the file since my JSON utilities didn't maintain hash order. Just something that may save you some time.

There is now [a kalite-es module up on the dev server](http://dev.worldpossible.org/cgi/viewmod.pl?module_id=58), though it requires a few manual edits to the configuration to work. But you can peek at that if you want to create a similar Swahili version...

**Comment by spatiald** [comment]
@needlestack Could you provide an example of the issues you saw with your install of the Spanish content for KA Lite?  I have it installed on the CAP with no apparent issues.  

You can check out a demo of the main page [here](http://rachel.golearn.us:85).  I "borrowed" your KA Lite module index.htmlf and rachel-index.php files. :)

You can check out a demo of the KA Lite in Spanish [here](http://rachel.golearn.us:86).

**Comment by needlestack** [comment]
So yeah - first thing I should say is that this is all sort of irrellevant as 0.16 was just released over the weekend and seems to fix most of the problems I was having (and is sure to introduce some new but hopefully minor issues, as updates often do!)

But to explain - there were a lot of videos (over 200) that appear as English in the 0.15 kalite. [For example](http://rachel.golearn.us:86/learn/khan/math/early-math/cc-early-math-counting-topic/cc-early-math-numbers-120/numbers-to-120/).

If you don't download the English videos before downloading the Spanish videos, these will just show up as unavailable. That was the issue that Jeremy reported above. If you install the English videos then it'll just show in English. My updates were to instead allow the new Spanish videos to be used.

The other issue was with disk caching - not sure if you enabled it on the CAP (it's off by default)... but when I enabled it it caused some parts of the site to be non-translated. And it also prevented me from modifying the topic tree (though that was me poking at the innards, so I voided the warranty :)) On the Pi if you don't enable caching, kalite startup takes around 10 minutes.

Anyway, with 0.16, all this is in the past :)
